### PR TITLE
New object detection model: Feature Pyramid Detector (FPN) - PascalVOC; improve mAP calculations

### PR DESCRIPTION
**Pull request by seldauyanik-maxim**

### CODE REVIEW

File: notebooks/DatasetVisualization_pascalvoc.ipynb [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    "import sys\n",
+    "\n",
+    "import numpy as np\n",
+    "import torch\n",
+    "\n",
+    "import matplotlib.patches as patches\n",
+    "import matplotlib.pyplot as plt\n",
+    "\n",
+    "sys.path.append(os.path.join(os.getcwd(), '..'))\n",
+    "sys.path.append(os.path.join(os.getcwd(), '../models/'))\n",
```
Comment (rotx-maxim) [inline notebooks/DatasetVisualization_pascalvoc.ipynb:28]: Please make this OS independent (os.path.join)

File: notebooks/DatasetVisualization_pascalvoc.ipynb [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
...NbQdx2liivZU0a8YpJWyzyuQhBoOKXEOA5T26jCZPdP4SKRIiZbVs+qitPTR7SbFftL6e4ax5GubdHWUFWVtJ3PZuzvH7C3t8ebb75JIR9XdU3diNKlsTVNM0MZA1pzenYqolbAfFbjxxGfW9Kdc2il2Vsu8wYBMUjAWZbrEJMoc2rNar1GobBVxXK5zzCOWGOYzWacF3FApQgxUDcSHI9Djx+dbNbOsbe3T13VVJXFhZDnfWToOkY3kGJgVtU58VFoa7k6h5tLw8uHDc+ogRmOKnnEcE0Iq2hRklRKCcEzl0OFhGFyAK2JStpCQ0yMLkztvKiEH6WVd3C5/DGM9MNA2/fMmhlaGzab1bazTikeXawYhwE3DlgNpqrBVty9aHnzgwc8uOiI8wP2j65ha1E4df3A9RT5lS989tLafn15kxNts3njkvn+VY6v3+KVj32cZ5+7Td+teXDvff7ZP/2sJGTGMo6eWbWkMg06GPZffImPfPzj/N5//9/n7PyUL79/h89/6R1efuklnrt9i+tXjgWFaHuG0eFDFJIxmSc0OnzvcO3AcHHB3fff4cGddzm782Vi+xBCS20VMfk8TkLmEeXEL22TGJAyfOGObcuwRfejoA3bwGI3eIgxbkULH9vqtdZZ0jxlN29BuKP3qNwCXdc1YULUI259yvn5+VeVdijHb7lXyP+vHsE5oq2l5BCkUyJGT2UrKltjTJUXQ2H4x0t932KfJXoHFl01aCPCKEX4pTwYVEnS4rZ0kCGqlP+PCYVPVJVFG4WP4h0gCnY7kcIEVm5FVKZD6hSX4HohBrGNjkmXBtzjtTmlNFqlLVs+pi3ikQdlgZGnel02ZfPOS4ts1l/Q2lBpm0liQkCs5gZlZuhmgUoaP4zEweHGET8OhOzh4DO0HdyA7g2NhcpE6tkBi7qhWSw5uHmbarnANDXGChoSnEMxMpuB1Q2VnaGVoq0aOm3wqzNictIibCx6vo+eLVnuHWLm+yIapYAYckAl1u4qRFQM2KxlUmDL5XxO08xp5guqeg/vREzJj54xRbxWBA1VMOjUUOsFo1lim4rZ0nLtWVAYrKnZ3LsLpwnfiVZFqBdUx9c4fvVjvPCxb2f/2k2Wh9dY7l+VLDDCRdvTD0KS67MDpSqiS8ZibEVlK5TLQm5GcWU546UXnuXWrZvsL2acn54SRtENiV7KMqQgH
```
Comment (rotx-maxim) [inline notebooks/DatasetVisualization_pascalvoc.ipynb:275]: path join

**Review by rotx-maxim — CHANGES_REQUESTED** [review]
Need os.path.join in Jupyter notebooks

File: models/ai87-fpndetector.py [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+
+    def forward(self, input_data):
+        """
+        Forward propagation for the Regression Module
+        """
+        out = self.res_conv1(input_data)
+        out = self.res_conv2(out)
+
+        out = self.conv5(out)
+
```
Comment (ermanok) [inline models/ai87-fpndetector.py:280]: unnecessary empty lines can be removed.

**Review by ermanok — DISMISSED** [review]
Looks great. I just have some minor comments about code refactor.

File: models/ai87-fpndetector.py [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+    def forward(self, input_data):
+        """
+        Forward propagation for the Classification Module
+        """
+        out = self.res_conv1(input_data)
+        out = self.res_conv2(out)
+
+        out = self.conv5(out)
+
+        out1 = out.permute(0, 2, 3, 1)
```
[the remark this answers is not in the document]
Comment (seldauyanik-maxim) [reply models/ai87-fpndetector.py]: Done - in last commit

File: models/ai87-fpndetector.py (hunk 3 above)
Comment (seldauyanik-maxim) [reply models/ai87-fpndetector.py:278]: Done - in last commit

**Review by ermanok — APPROVED** [review]
Looks good.

File: notebooks/DatasetVisualization_pascalvoc.ipynb [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+      "Train dataset length: 16551\n",
+      "\n",
+      "Test dataset length: 4952\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "device = \"cpu\"\n",
+    "data_path = '/data_ssd/'\n",
```
Comment (rotx-maxim) [inline notebooks/DatasetVisualization_pascalvoc.ipynb:58]: We should mark this as "CHANGE ME TO DATA DIRECTORY", remove at least the trailing slash and then use os.path.join later

File: notebooks/TinySSDModelVisualization_svhn.ipynb [hunk 6]
```diff
@@ -52,7 +58,7 @@
    ],
    "source": [
     "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
-    "data_path = '/data/ml/'\n",
+    "data_path = '/data_ssd/'\n",
```
Comment (rotx-maxim) [inline notebooks/TinySSDModelVisualization_svhn.ipynb:61]: mark as "CHANGE ME", remove trailing slash, use with os.path.join

File: notebooks/TinySSDModelVisualization_svhn.ipynb [hunk 7]
```diff
@@ -73,34 +79,32 @@
...
      "text": [
+      "{'start_epoch': 25, 'weight_bits': 8, 'shift_quantile': 0.995}\n",
       "Configuring device: MAX78000, simulate=False.\n"
      ]
     }
    ],
    "source": [
     "num_classes = 11\n",
     "\n",
+    "qat_yaml_file_used_in_training = '../policies/qat_policy_svhn.yaml'\n",
```
Comment (rotx-maxim) [inline notebooks/TinySSDModelVisualization_svhn.ipynb:90]: os.path.join

**Review by rotx-maxim — CHANGES_REQUESTED** [review]
There are more instances of Unix paths in the notebooks. Where possible, we should change those to work on Windows as well (os.path.join)

File: datasets/pascalvoc.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+                self.dataset = torch.utils.data.ConcatDataset([self.dataset, dataset_temp])
+            else:
+                self.dataset = dataset_temp
+
+    def __transforms_func(self, image, target):
+        """
+        Torch vision VOCDetection dataset's __get_item__ returns: (image, target)
+        where target is a dictionary of the XML tree.
+        This transform function applies transformations on:
+            * images: Normalized, resized, ai8xized
```
Comment (ermanok) [inline datasets/pascalvoc.py:98]: the image should not be ai8xized at this level. This transformation is applied at the very end; in __get_item method.

File: models/ai87-fpndetector.py [hunk 9]
```diff
@@ -0,0 +1,544 @@
...
+                image_labels = image_labels[sort_ind][:top_k]  # (top_k)
+
+            # Append to lists that store predicted boxes and scores for all images
+            all_images_boxes.append(image_boxes)
+            all_images_labels.append(image_labels)
+            all_images_scores.append(image_scores)
+
+        return all_images_boxes, all_images_labels, all_images_scores  # lists of length batch_size
+
+    def are_locations_wide(self):
```
Comment (oguzhanbsolak) [inline models/ai87-fpndetector.py:517]: Should we modify tinierssd and tinierssd-face model for adding are_locations_wide and are_scores_wide functions with this PR?

Comment (seldauyanik-maxim) [reply models/ai87-fpndetector.py:517]: Train.py checks those functions only if they exist and if not treats all outputs as no-wide. Therefore tinierssd and tinierssd-face models will be treated as expected (as their outputs are non-wide). However, I think adding those would be more readable even if not functionally necessary

File: notebooks/TinySSDModelVisualization_svhn.ipynb (hunk 7 above)
Comment (seldauyanik-maxim) [reply notebooks/TinySSDModelVisualization_svhn.ipynb]: Sure, done in the last commit

File: datasets/pascalvoc.py (hunk 8 above)
Comment (seldauyanik-maxim) [reply datasets/pascalvoc.py]: Correct, the comments are revised in the last commit

**Review by rotx-maxim — CHANGES_REQUESTED** [review]
Please run dos2unix on
   models/ai85net-tinierssd-face.py
   utils/object_detection_utils.py

**Pull request merged**